File: Grind75/23-MaxDepth/cpp/MaxDepth.cpp
#include "MaxDepth.h"
#include <Tree.h>
#include <stack>
#include <utility>

using namespace BST;

int MaxDepth::maxDepth(TreeNode* root) {
  if (!root)
    return 0;
  int leftDepth = maxDepth(root->left);
  int rightDepth = maxDepth(root->right);
  return (leftDepth > rightDepth ? leftDepth : rightDepth) + 1;
}

int MaxDepth::maxDepth2(TreeNode *root) {
  if (!root)
    return 0;

  std::stack<std::pair<TreeNode*, int>> s;
  s.push(std::make_pair(root, 1));
  int maxDepth = 0;

  while (!s.empty()) {
    std::pair<TreeNode *, int> t = s.top();
    s.pop();
    TreeNode *node = t.first;
    int depth = t.second;
    if (depth > maxDepth)
      maxDepth = depth;
    if (node->left)
      s.push(std::make_pair(node->left, depth+1));
    if (node->right)
      s.push(std::make_pair(node->right, depth+1));
  }
  return maxDepth;
}

File: Grind75/23-MaxDepth/python/MaxDepth.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from Tree import *

class Solution(object):
  def maxDepth(self, root: Optional[TreeNode]) -> int:
    if root is None:
      return 0
    return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1


if __name__ == '__main__':
  soln = Solution()
  testCases: list[tuple[list, int]] = [
    ([], 0),
    ([1], 1),
    ([2,1], 2),
    ([2,1,3], 2),
    ([2,1,3,4], 3),
    ([2,1,7,5,4,3,6], 5),
  ]

  for lst, depth in testCases:
    tree = makeBST(lst)
    printTree(tree)
    print(f'should have a depth of {depth}, ', end='')
    res = soln.maxDepth(tree)
    print(f'got {res}')
    assert res == depth
  print('TESTS PASSED!')

